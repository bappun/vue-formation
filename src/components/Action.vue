<template>
    <icon tag="a" :href="value" v-if="action === 'edit'" class="text-info icon-md"><PenIcon/></icon>
</template>

<script>
import PenIcon from '@static/icons/pen.svg'
import Icon from './Icon.vue'

export default {
  name: 'Action',
  components: { Icon, PenIcon },
  props: {
    action: String,
    value: String
  }
}
</script>
