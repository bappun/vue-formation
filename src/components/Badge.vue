<template>
    <span class="badge badge-pill" :class="cssClass" v-text="text"></span>
</template>

<script>
export default {
  name: 'Badge',
  props: {
    value: [String, Array]
  },
  data () {
    const [text, cssClass] = Array.isArray(this.value) ? this.value : [this.value, 'badge-secondary']
    return { text, cssClass }
  }
}
</script>
