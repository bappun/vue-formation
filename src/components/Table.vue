<template>
    <Loader v-if="isLoading"/>
    <div v-else-if="sortedValues">
        <div v-if="sortedValues.length <= 0" class="alert alert-info">No data found</div>
        <div v-else class="table-responsive bg-white">
            <table class="table table-bordered table-hover">
                <thead class="thead-light">
                <tr>
                    <th
                        :tabindex="entry.isSortable ? 0 : undefined"
                        :key="entry.name" v-for="entry in config"
                        @keyup.space="sortBy(entry)"
                        @click="sortBy(entry)"
                        :class="{'cursor-pointer' : entry.isSortable}">
                            <span>{{ entry.text }}</span>
                            <icon
                                v-if="entry.isSortable"
                                :class="{'active': entry.name === sortKey, 'rotate-2': sortOrders[entry.name] > 0}">
                                  <ArrowIcon/>
                            </icon>
                    </th>
                </tr>
                </thead>
                <transition-group tag="tbody">
                    <tr :key="row.id || rIndex" v-for="(row, rIndex) in sortedValues">
                        <td :key="col.name || cIndex" v-for="(col, cIndex) in config" :class="col.extraClasses">
                            <component v-if="col.component" :is="col.component" v-bind="{ value: row[col.name], ...col.props }"/>
                            <template v-else>{{ row[col.name] }}</template>
                        </td>
                    </tr>
                </transition-group>
            </table>
        </div>
    </div>
</template>

<script>
import Loader from '@frontend/components/Loader.vue'

const Icon = () => import('@frontend/components/Icon.vue')
const ArrowIcon = () => import('@static/icons/arrow-up.svg')

export default {
  name: 'Table',
  components: { Loader, Icon, ArrowIcon },
  props: {
    values: Array,
    config: Array,
    isLoading: Boolean
  },
  data () {
    const sortOrders = {}
    this.config.forEach((entry) => {
      if (entry.isSortable) sortOrders[entry.name] = -1
    })

    return {
      sortKey: '',
      sortOrders
    }
  },
  computed: {
    sortedValues () {
      const { sortKey, sortOrders } = this
      let { values } = this
      const order = sortOrders[sortKey] || 1
      if (sortKey) {
        values = values.slice().sort((a, b) => {
          a = a[sortKey] // eslint-disable-line no-param-reassign
          b = b[sortKey] // eslint-disable-line no-param-reassign
          return (a === b ? 0 : a > b ? 1 : -1) * order // eslint-disable-line no-nested-ternary
        })
      }
      return values
    }
  },
  methods: {
    sortBy (entry) {
      if (entry.isSortable) {
        const key = entry.name
        this.sortKey = key
        this.sortOrders[key] = this.sortOrders[key] * -1
      }
    }
  }
}
</script>
