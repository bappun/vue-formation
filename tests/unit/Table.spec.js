import { shallowMount } from '@vue/test-utils'
import Table from '@frontend/components/Table.vue'
import Action from '@frontend/components/Action'
import Badge from '@frontend/components/Badge'
import SyncIcon from '@static/icons/sync.svg'
import SyncCheckIcon from '@static/icons/sync_check.svg'
import members from '@fixtures/members.json'

const config = [
  {
    text: 'Last name',
    name: 'last_name',
    isSortable: true
  },
  {
    text: 'First name',
    name: 'first_name',
    isSortable: true
  },
  {
    text: 'Email',
    name: 'email',
    isSortable: true
  },
  {
    text: 'Language',
    name: 'language',
    isSortable: true,
    extraClasses: ['text-uppercase'],
    component: Badge
  },
  {
    text: 'State',
    name: 'state_label',
    isSortable: true,
    component: Badge
  },
  {
    text: 'Qualtrics',
    name: 'qualtrics',
    isSortable: true,
    component: Boolean,
    props: {
      icons: [SyncIcon, SyncCheckIcon]
    }
  },
  {
    text: 'Edit',
    name: 'edit',
    component: Action,
    props: {
      action: 'edit'
    }
  }
]

describe('Table.vue', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(Table, {
      propsData: { values: members, config }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('header click', () => {
    const wrapper = shallowMount(Table, {
      propsData: { values: members, config }
    })
    const el = wrapper.find('th:first-child')
    el.trigger('click')
    expect(el.html()).toContain('active rotate-2')
  })
})
