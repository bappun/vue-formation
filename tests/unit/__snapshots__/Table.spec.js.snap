// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.vue renders correctly 1`] = `
<div>
  <div
    class="table-responsive bg-white"
  >
    <table
      class="table table-bordered table-hover"
    >
      <thead
        class="thead-light"
      >
        <tr>
          <th
            class="cursor-pointer"
            tabindex="0"
          >
            <span>
              Last name
            </span>
             
            <!---->
          </th>
          <th
            class="cursor-pointer"
            tabindex="0"
          >
            <span>
              First name
            </span>
             
            <!---->
          </th>
          <th
            class="cursor-pointer"
            tabindex="0"
          >
            <span>
              Email
            </span>
             
            <!---->
          </th>
          <th
            class="cursor-pointer"
            tabindex="0"
          >
            <span>
              Language
            </span>
             
            <!---->
          </th>
          <th
            class="cursor-pointer"
            tabindex="0"
          >
            <span>
              State
            </span>
             
            <!---->
          </th>
          <th
            class="cursor-pointer"
            tabindex="0"
          >
            <span>
              Qualtrics
            </span>
             
            <!---->
          </th>
          <th
            class=""
          >
            <span>
              Edit
            </span>
             
            <!---->
          </th>
        </tr>
      </thead>
       
      <span
        tag="tbody"
      >
        <tr>
          <td>
            First
          </td>
          <td>
            John
          </td>
          <td>
            John.first@example.com
          </td>
          <td
            class="text-uppercase"
          >
            <badge-stub
              value="FR"
            />
          </td>
          <td>
            <badge-stub
              value="Not notified,badge-warning"
            />
          </td>
          <td>
            <!---->
          </td>
          <td>
            <action-stub
              action="edit"
              value="/m/members/1/update/"
            />
          </td>
        </tr>
        <tr>
          <td>
            Second
          </td>
          <td>
            John
          </td>
          <td>
            John.second@example.com
          </td>
          <td
            class="text-uppercase"
          >
            <badge-stub
              value="EN"
            />
          </td>
          <td>
            <badge-stub
              value="Active,badge-success"
            />
          </td>
          <td>
            <!---->
          </td>
          <td>
            <action-stub
              action="edit"
              value="/m/members/2/update/"
            />
          </td>
        </tr>
        <tr>
          <td>
            Third
          </td>
          <td>
            John
          </td>
          <td>
            John.third@example.com
          </td>
          <td
            class="text-uppercase"
          >
            <badge-stub
              value="EN"
            />
          </td>
          <td>
            <badge-stub
              value="Active,badge-success"
            />
          </td>
          <td>
            <!---->
          </td>
          <td>
            <action-stub
              action="edit"
              value="/m/members/3/update/"
            />
          </td>
        </tr>
      </span>
    </table>
  </div>
</div>
`;
